### PR TITLE
Cleanup PrunedLiveBlocks

### PR DESCRIPTION
**Pull request by atrick**
Use BasicBlockBitfield to record per-block liveness state. This has
been the intention since BasicBlockBitfield was first introduced.

    Remove the per-field bitfield from PrunedLiveBlocks. This
    (re)specializes the data structure for scalar liveness and drastically
    simplifies the implementation.

    This utility is fundamental to all ownership utilities. It will be on
    the critical path in many areas of the compiler, including at
    -Onone. It needs to be minimal and as easy as possible for compiler
    engineers to understand, investigate, and debug.

    This is in preparation for fixing bugs related to multi-def liveness
    as used by the move checker.

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
CI failure:
```
Library.swiftinterface:79:16: error: type 'SPIAvailableStruct.UnavailableNestedActor' does not conform to protocol 'Actor'
  public actor UnavailableNestedActor {
```

**Comment by atrick** [comment]
@swift-ci smoke test macos platform

File: include/swift/SIL/PrunedLiveness.h [hunk 1]
```diff
@@ -181,210 +173,101 @@ class PrunedLiveBlocks {
...
-  SmallVectorImpl<SILBasicBlock *> *discoveredBlocks;
+  SmallVectorImpl<SILBasicBlock *> *discoveredBlocks = nullptr;
+
+  /// Only a clean bitfield can be initialized.
+  SWIFT_ASSERT_ONLY_DECL(bool cleanFlag = true);
 
-  /// Once the first use has been seen, no definitions can be added.
-  SWIFT_ASSERT_ONLY_DECL(bool seenUse = false);
+  /// Once the first def has been initialized, uses can be added.
+  SWIFT_ASSERT_ONLY_DECL(bool initializedFlag = false);
```
Comment (drodriguez) [inline include/swift/SIL/PrunedLiveness.h:197]: Is this is a `SWIFT_ASSERT_ONLY_DECL`, do the other mentions like `bool isInitialized()` need to be in similar blocks?

File: include/swift/SIL/PrunedLiveness.h [hunk 2]
```diff
@@ -181,210 +173,101 @@ class PrunedLiveBlocks {
...
-  unsigned numBitsToTrack;
+  /// Map all blocks to an IsLive state.
+  BasicBlockBitfield liveBlocks;
 
   /// Optional vector of live blocks for clients that deterministically iterate.
-  SmallVectorImpl<SILBasicBlock *> *discoveredBlocks;
+  SmallVectorImpl<SILBasicBlock *> *discoveredBlocks = nullptr;
+
+  /// Only a clean bitfield can be initialized.
+  SWIFT_ASSERT_ONLY_DECL(bool cleanFlag = true);
```
Comment (drodriguez) [inline include/swift/SIL/PrunedLiveness.h:194]: Same with this one. There's usages, for example, in `invalidate()`.

File: include/swift/SIL/PrunedLiveness.h (hunk 1 above)
Comment (atrick) [reply include/swift/SIL/PrunedLiveness.h:197]: @drodriguez yes, thank you. That was an oversight.
https://github.com/apple/swift/pull/64581

Comment (atrick) [reply include/swift/SIL/PrunedLiveness.h:197]: @drodriguez I'm curious how you noticed this. Did the release build fail to compile? I thought full PR testing covered that.

Comment (drodriguez) [reply include/swift/SIL/PrunedLiveness.h:197]: We build nightly both asserts and non-asserts for the compiler and run the tests. It catches many of these kind of errors that ci.swift.org does not catch (a lot of tests for experimental features should be `REQUIRES: asserts`, but aren't, for example).

**Pull request merged**